### ISSUE TITLE
`processObliv.ml` cannot handle C labels, like `__gmp_done` in <gmp.h>

### ISSUE THREAD

**Issue by weikengchen** [samee/obliv-c#59]
This is relevant to https://github.com/samee/obliv-c/issues/58.

## Background
The issue shows that labels are discarded magically, not by CIL, but by Obliv-C's modification.

We know `oblivcc` processes all C/OC files with `cilly`. The main change from `CIL` to `Obliv-C` backend is `processObliv.ml`. After the test,  we find that the label is discarded by`processObliv.ml`.

We know for sure that it is not practical to consider `goto` in an `obliv if`, but in non-obliv areas, people may import headers from other libraries. 

Although `goto` is strongly not encouraged nowadays for daily programming, some third-party libraries, like GMP (<gmp.h>) uses `goto` -- this will cause compile errors in Obliv-C.

## Reason
My first time OCaml :)

After a careful debug, we find the following:

`codegenVisitor` rewrites all statements. When it matches `Instr ilist` where `ilist` should be a list, it calls `codegenInstr` to rewrite. This one does not produce a successful new statement for C labels, in which `List.length ilist` is 0 and you cannot reproduce a statement from `ilist`. 

```
    | Instr ilist -> 
        let nestedGen = codegenInstr curCond tmpVar isDeepVar in
        ChangeTo (mkStmt (Instr (mapcat nestedGen ilist)))
```

Very special the way CIL treats labels. The `vstmt` on the labels also has no line number and filename associated with it when we use `E.warn` to debug. 

## Solution
It can be fixed by `DoChildren` instead of `ChangeTo` when `ilist` is empty.
```
    | Instr ilist -> 
        let num = List.length ilist in
        let nestedGen = codegenInstr curCond tmpVar isDeepVar in
        if num == 0 then DoChildren
        else ChangeTo (mkStmt (Instr (mapcat nestedGen ilist)))
```

(A fixed version of processObliv.ml: https://gist.github.com/weikengchen/ed7ec649fbac17b67df5cce7abf60194)

## Side note on `goto` in `obliv` areas
Should we prohibit `goto` in `obliv`?

It will be adding in `class controlCheckVisitor` with `gotoOk` similar to `returnOk`, with the following `E.error`:

```
    | Goto (_,loc) -> if not !gotoOk then
             E.s (E.error "%a: unexpected goto" d_loc loc)
```

## Question
need a PR?

**Comment by samee** [comment]
I don't understand why that change would fix anything. Can you explain your reasoning a bit?

I was going to just disallow `goto` altogether for `.oc` files. We shouldn't need to import `gmp.h` from anything that gets included into `.oc`.

Unfortunately, our frontend actually *introduces* new goto statements whenever we encounter logical operators `&&`, `||`: https://github.com/samee/obliv-c/blob/obliv-c/src/frontc/cabs2cil.ml#L4869

We must be careful not to disallow or adversely affect those.

**Comment by weikengchen** [comment]
(This note has been corrected.
Previously, this note said ``killing goto``, they are corrected as ``killing labels``.
Sorry for making misunderstanding.)

## Background

My application uses a reactive protocol. The protocol calls another library, which rely on GMP.

```
xxx.c
xxx.h
xxx.oc   <-   libyyy    <-   gmp.h
```

If I put all those function definitions into the `.c` instead of `.oc`. The protocol will need to call the functions defined in `.c`. We need to build a bridge.

```
xxx.c.     <-   libyyy    <-   gmp.h
     -> libyyy_interface
xxx.h
     <- defined libyyy_interface
xxx.oc   
     <- use libyyy_interface, repetitively
```

And we need to avoid using any GMP data structures in `xxx.oc`.
This adds implementations overhead. 

## Is fully disallowing `goto` necessary?

The example above shows that, if you allow including `<gmp.h>`, which has `goto` somewhere, you do not need additional implementation. It is just convenient.

Not every `goto` breaks the security. Some `goto` may be fine.
And when I use a third-party library, like `<gmp.h>`, I cannot change it too much.

And I kind of prefer not to distinguish `.oc` and `.c` too much. Just feel not necessary. 
They finally go to the same binary. 

## Why previously `goto` fails?

This is about the issue #58.
As you mentioned, `As for the rationale, it was because the obliv-related type rules are hard to design for goto, so we decided not to support it at all. We should probably just raise an error instead of silently generating the wrong code though.`

Currently, `processObliv` does not explicitly forbid `goto`. But some labels for `goto` is just dropped.

The reason is that there is an implementational detail in `codegenVisitor`, which drops some labels because some labels are statements with an empty list and cannot be reconstructed. So when `codegenVisitor` rewrites, some of these labels disappears.

I feel it is not necessary to drop these labels. 
We know that jumping into an `obliv if` or jumping out of an `obliv if` is problematic, and `label` inside `obliv if` is problematic, but I feel it is the programmer to be responsible to use `goto` correctly. And anyway, we should not drop labels.

## How this fix works?

This fix changes `codegenVisitor`, where some labels are dropped.
Now these labels are not dropped.

## Summary

So, in summary, it is GMP header which has a label.
I will not use that label. I use low-level functions of GMP directly.
But I want to include `<gmp.h>` for data structure definitions.

I can include `<gmp.h>` in `.c` rather than `.oc`, but I just feel not necessary.

However, I cannot include `<gmp.h>` in `.oc` because `processObliv` drops the label and cause a gcc compiler error.

I think "dropping labels" is a bug, rather than a feature. So I feel we should not drop labels. Some are not in `obliv if` and will not cause security problems. They are jumping outside and I might even not use them at all. They just come with the header file of `gmplib`. 

## Suggestion

Keep these labels.

**Comment by weikengchen** [comment]
I will do some experiments on whether `&&` or `||` used in `obliv if` generates new goto that are likely to be forbidden by a check like
```
    | Goto (_,loc) -> if not !gotoOk then
             E.s (E.error "%a: unexpected goto" d_loc loc)
```
if I have time :)

**Comment by samee** [comment]
Regarding separation, you may be able to use `#ifdef __oblivious_c` to check  if you are currently being compiled by oblivcc or gcc, and use that to decide whether gmp.h should be included. But yes, in general we prefer if code separates out the C-specific, Obliv-C-specific, and portable headers. The portable parts should be includable in both.

**Comment by weikengchen** [comment]
Samee, do you know any test example that triggers the following:
```
Unfortunately, our frontend actually introduces new goto 
statements whenever we encounter logical operators &&, 
||: https://github.com/samee/obliv-c/blob/obliv-c/src/frontc/cabs2cil.ml#L4869
```

`Cilly` currently makes,
```
      if(a == 1 && b == 2){
             xxx
      }
```
into 
```
     if(a == 1){
            if(b == 2){
                   xxx
            }
     }
```

and not generating `goto` for the test examples I tried.

I think a debug will be needed here. It is possible that `processObliv` discards the label generated by L4869 due to the problem I mentioned above.

(Confirmed. See below)

**Comment by weikengchen** [comment]
Got one: https://people.eecs.berkeley.edu/~necula/cil/cil004.html
Result: Goto kept. But label disappears.

See the next note for details.

**Comment by weikengchen** [comment]
@samee 

Now there is a stronger reason why we need to fix.

```
int main(){
  int x;
  int y = x ? 2 : 4;
  int z = x || y;
  // Here we duplicate the return statement
  if(x && y) { return 0; } else { return 1; }
  // To avoid excessive duplication, CIL uses goto's for 
  // statement that have more than 5 instructions
  if(x && y || z) { x ++; y ++; z ++; x ++; y ++; return z; }
}
```

Without `processObliv`, it generates:
```
#line 9
  if (x) {
#line 9
    if (y) {
#line 9
      goto _L;
    } else {
#line 9
      goto _L___0;
    }
  } else
  _L___0: /* CIL Label */ 
#line 9
  if (z) {
    _L: /* CIL Label */ 
#line 9
    xxxxxxxxx
#line 9
    return (z);
  }
```

However, `processObliv` kills the _L label.

```
#line 9
  if ((_Bool )x) {
#line 9
    if ((_Bool )y) {
#line 9
      goto _L;
    } else {
#line 9
      goto _L___0;
    }
  } else
  _L___0: /* CIL Label */
#line 9
  if ((_Bool )z) {
#line 9
    __cil_tmp6 = (int )1;
     xxxxxxx
#line 9
    return (z);
  }
```

causing a `gcc` error:
```
test.c: In function ‘main’:
test.c:9:7: error: label ‘_L’ used but not defined
   if(x && y || z) { x ++; y ++; z ++; x ++; y ++; return z; }
```

**Comment by weikengchen** [comment]
And a correction:

I said `processObliv` kills `goto`.

It was wrong. It should be:
`processObliv` kills labels that `goto` uses.

Sorry for making misunderstanding. The note was already ~10 days ago and I forgot most of them.

**Comment by weikengchen** [comment]
And regarding the `goto` check,

You are right. 
If the CIL adds these `goto`, they may reside in `obliv if`. A check similar to `return` and `switch` can cause false alarms.

Two options:
(1) If this is the only place that CIL adds `goto` and `label`, we can see if we can label these `goto`, like adding them into the whitelist.
(2) Let's forget the `goto` check.

Limitation of (1):
- If we restrict all `goto`, we may encounter the same trouble like `<gmp.h>`.
- If we only forbid `goto` in `obliv if`, it is not accurate. Because someone can use `goto` to jump from outside `obliv if` into an `obliv if`.

**Comment by samee** [comment]
Here's a possibly cleaner fix:

https://github.com/samee/obliv-c/blob/obliv-c/src/frontc/cparser.mly#L920-L923

Let's have both these lines produce error. This happens long before any preprocessing, right when input source is being parsed. This makes sure we are not dealing with any newly generated goto, only the ones directly present in source. Once we fix this, we can decide what to do about logical operators.

**Comment by weikengchen** [comment]
I would like to mention again that the error is caused by deleted labels, not deleted goto.

**Comment by samee** [comment]
I definitely prefer not to support including arbitrary C libraries in Obliv-C. C is a complex beast to process consistently as it is, so supporting a well-defined subset is always a good strategy for not introducing unintentional inconsistencies. That said, I cannot find any concrete problems with your suggestion.

In fact, I do see the value in supporting `goto` for some cases. For example, since C doesn't have destructors or exceptions, there is no way to do complicated error handling other than by using `goto`. But we do have to make sure that the `goto` label and `goto` statement never crosses `obliv if` boundaries. For example, the following should be forbidden:

        obliv if (...) {
            goto foo;
        }
        ...
    foo:

Similarly, the converse where a `goto` label lands inside a new `obliv if` scope from the outside should also be forbidden. If we can do this cleanly and consistently, while taking care of all possible corner cases, I can accept that change. But in general, I always err on the side of not having a feature as opposed to having one that is inconsistent.

In case it helps, we already have very similar rules concerning `break`, `continue`, and `return`, in that they should not cross any `obliv if` boundaries.

**Comment by weikengchen** [comment]
I agree that we would hold off on the `goto` limitation.

I will close this issue, and make a separated one for my main concern.

**Comment by weikengchen** [comment]
Sorry, I should have made this one more clear.
